services/vision/obstacledepth/obstacle_depth.go

```diff
@@ -0,0 +1,376 @@
+	"go.viam.com/rdk/rimage"
+	"go.viam.com/rdk/rimage/transform"
+	"go.viam.com/rdk/robot"
+	svision "go.viam.com/rdk/services/vision"
+	"go.viam.com/rdk/spatialmath"
+	"go.viam.com/rdk/utils"
+	vision "go.viam.com/rdk/vision"
+	"go.viam.com/rdk/vision/segmentation"
+)
+
+var model = resource.DefaultModelFamily.WithModel("obstacle_depth")
+
+// ObstaclesDepthConfig specifies the parameters to be used for the obstacle depth service.
+type ObstaclesDepthConfig struct {
+	K          int     `json:"k"`
+	Hmin       float64 `json:"hmin"`
+	Hmax       float64 `json:"hmax"`
+	ThetaMax   float64 `json:"theta_max"`
+	ReturnPCDs bool    `json:"return_pcds"`
+	intrinsics *transform.PinholeCameraIntrinsics
+}
+
+// obsPoint is a useful intermediate struct (which pts are obstacles).
+type obsPoint struct {
+	point      image.Point
+	isObstacle bool
+	depth      rimage.Depth
+}
+
+// obsDepth is the underlying struct actually used by the service.
+type obsDepth struct {
+	dm         *rimage.DepthMap
+	ptChunks   [][]obsPoint
+	ptList     []obsPoint
+	hMin       float64
+	hMax       float64
+	sinTheta   float64
+	intrinsics *transform.PinholeCameraIntrinsics
+	returnPCDs bool
+	k          int
+}
+
+const (
+	// the first 3 consts are parameters from Manduchi et al.
+	defaultHmin     = 0.0
+	defaultHmax     = 150.0
+	defaultThetamax = math.Pi / 4
+	// the last 3 consts are hyperparameters that can be tweaked for performance improvement.
+	chunkSize  = 200 // we send chunkSize points in each goroutine
+	sampleN    = 4   // we sample 1 in every sampleN depth points
+	maxThreads = 300 // we will run at most maxThreads goroutines
+)
+
+func init() {
+	resource.RegisterService(svision.API, model, resource.Registration[svision.Service, *ObstaclesDepthConfig]{
+		DeprecatedRobotConstructor: func(ctx context.Context, r any, c resource.Config, logger golog.Logger) (svision.Service, error) {
+			attrs, err := resource.NativeConfig[*ObstaclesDepthConfig](c)
+			if err != nil {
+				return nil, err
+			}
+			actualR, err := utils.AssertType[robot.Robot](r)
+			if err != nil {
+				return nil, err
+			}
+			return registerObstacleDepth(ctx, c.ResourceName(), attrs, actualR)
+		},
+	})
+}
+
+// Validate ensures all parts of the config are valid.
+func (config *ObstaclesDepthConfig) Validate(path string) ([]string, error) {
+	deps := []string{}
+	if config.K < 1 || config.K > 50 {
+		return nil, errors.New("invalid K, pick an integer between 1 and 50 (10 recommended)")
+	}
+	if (config.Hmin >= config.Hmax) && config.Hmax != 0 {
+		return nil, errors.New("Hmin should be less than Hmax")
+	}
+	return deps, nil
+}
+
+func registerObstacleDepth(
+	ctx context.Context,
+	name resource.Name,
+	conf *ObstaclesDepthConfig,
+	r robot.Robot,
+) (svision.Service, error) {
+	_, span := trace.StartSpan(ctx, "service::vision::registerObstacleDistanceDetector")
+	defer span.End()
+	if conf == nil {
+		return nil, errors.New("config for obstacle_depth cannot be nil")
+	}
+
```

Comment:
The only thing I feel weird about here is that we can't (?) know if the camera they choose (src) has underlying intrinsics at **this** level.